system/security/digisign/digisign.cpp

```diff
@@ -0,0 +1,253 @@
+#endif
+    }
+
+    virtual ~CDigitalSignatureManager()
+    {
+    }
+
+    bool isDigiSignerConfigured()
+    {
+        return signingConfigured;
+    }
+
+    bool isDigiVerifierConfigured()
+    {
+        return verifyingConfigured;
+    }
+
+    bool digiInit(bool isSigning, const char * keyBuff, const char * passphrase, EVP_MD_CTX * * ctx, EVP_PKEY * * PKey)
+    {
+        //create an RSA object from public key
+        BIO * keybio = BIO_new_mem_buf((void*) keyBuff, -1);
+        if (nullptr == keybio)
+        {
+            EVP_THROW("digiSign:BIO_new_mem_buf: %s");
+        }
+
+        RSA * rsa;
+        if (isSigning)
+            rsa = PEM_read_bio_RSAPrivateKey(keybio, nullptr, nullptr, (void*)passphrase);
+        else
+            rsa = PEM_read_bio_RSA_PUBKEY(keybio, nullptr, nullptr, nullptr);
+        BIO_free_all(keybio);
+        if (nullptr == rsa)
+        {
+            if (isSigning)
+                EVP_THROW("digiSign:PEM_read_bio_RSAPrivateKey: %s")
+            else
+                EVP_THROW("digiSign:PEM_read_bio_RSA_PUBKEY: %s")
+        }
+
+        EVP_PKEY* pKey = EVP_PKEY_new();
+        if (nullptr == pKey)
+        {
+            EVP_THROW("digiSign:EVP_PKEY_new: %s");
+        }
+        EVP_PKEY_assign_RSA(pKey, rsa);//take ownership of the rsa. pKey will free rsa
+
+        EVP_MD_CTX * RSACtx = EVP_MD_CTX_create();//allocate, initializes and return a digest context
+        if (nullptr == RSACtx)
+        {
+            EVP_PKEY_free(pKey);
+            EVP_THROW("digiSign:EVP_MD_CTX_create: %s");
+        }
+
+        //initialize context for SHA-256 hashing function
+        int rc;
+        if (isSigning)
+            rc = EVP_DigestSignInit(RSACtx, nullptr, EVP_sha256(), nullptr, pKey);
+        else
+            rc = EVP_DigestVerifyInit(RSACtx, nullptr, EVP_sha256(), nullptr, pKey);
+        if (rc <= 0)
+        {
+            EVP_CLEANUP(pKey, RSACtx);//cleans allocated key and digest context
+            if (isSigning)
+                EVP_THROW("digiSign:EVP_DigestSignInit: %s")
+            else
+                EVP_THROW("digiSign:EVP_DigestVerifyInit: %s")
+        }
+        *ctx = RSACtx;
+        *PKey = pKey;
+        return true;
+    }
+
+    //Create base 64 encoded digital signature of given text string
+    bool digiSign(const char * text, StringBuffer & b64Signature)
+    {
+        if (!signingConfigured)
+            throw MakeStringException(-1, "digiSign:Creating Digital Signatures not configured");
+
+        if (privateKeyBuff.isEmpty())
+            privateKeyBuff.loadFile(privateKeyFile.str());
+        if (privateKeyBuff.isEmpty())
+            throw MakeStringException(-1, "digiSign:Cannot load private key file");
+
+        CriticalBlock b(digiSignCrit);
+
+#ifdef _USE_OPENSSL
+        EVP_MD_CTX * RSACtx;
+        EVP_PKEY * EVPKey;
+        digiInit(true, privateKeyBuff.str(), passphraseBuff.str(), &RSACtx, &EVPKey);
```

Comment:
Just a note, passphraseBuff from file may be encrypted and it is used here as is.  Perhaps a comment that passphraseBuff may need to be decrypted here before being used.